pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnSingleValueReaderWriter.java

```diff
@@ -17,131 +17,206 @@
-    writer.setBytes(row, 0, bytes);
+    addBufferIfNeeded(row);
+    int rowInChunk = getRowInChunk(row);
+    currentWriter.setBytes(rowInChunk, 0, bytes);
   }
 
   @Override
   public char getChar(int row) {
-    return reader.getChar(row, 0);
+    int rowInChunk = getRowInChunk(row);
+    int bufferId = getBufferId(row);
+    return readers.get(bufferId).getChar(rowInChunk, 0);
   }
 
   @Override
   public short getShort(int row) {
-    return reader.getShort(row, 0);
+    int rowInChunk = getRowInChunk(row);
+    int bufferId = getBufferId(row);
+    return readers.get(bufferId).getShort(rowInChunk, 0);
   }
 
   @Override
   public int getInt(int row) {
-    return reader.getInt(row, 0);
+    int rowInChunk = getRowInChunk(row);
+    int bufferId = getBufferId(row);
+    return readers.get(bufferId).getInt(rowInChunk, 0);
   }
 
   @Override
   public long getLong(int row) {
-    return reader.getLong(row, 0);
+    int rowInChunk = getRowInChunk(row);
+    int bufferId = getBufferId(row);
+    return readers.get(bufferId).getLong(rowInChunk, 0);
   }
 
   @Override
   public float getFloat(int row) {
-    return reader.getFloat(row, 0);
+    int rowInChunk = getRowInChunk(row);
+    int bufferId = getBufferId(row);
+    return readers.get(bufferId).getFloat(rowInChunk, 0);
   }
 
   @Override
   public double getDouble(int row) {
-    return reader.getDouble(row, 0);
+    int rowInChunk = getRowInChunk(row);
+    int bufferId = getBufferId(row);
+    return readers.get(bufferId).getDouble(rowInChunk, 0);
   }
 
   @Override
   public String getString(int row) {
-    return reader.getString(row, 0);
+    int rowInChunk = getRowInChunk(row);
+    int bufferId = getBufferId(row);
+    return readers.get(bufferId).getString(rowInChunk, 0);
   }
 
   @Override
   public byte[] getBytes(int row) {
-    return reader.getBytes(row, 0);
+    int rowInChunk = getRowInChunk(row);
+    int bufferId = getBufferId(row);
+    return readers.get(bufferId).getBytes(rowInChunk, 0);
   }
 
   @Override
   public void readValues(int[] rows, int rowStartPos, int rowSize, int[] values, int valuesStartPos) {
-    reader.readIntValues(rows, 0, rowStartPos, rowSize, values, valuesStartPos);
+    /*
+     * TODO
+     * If we assume that the row IDs in 'rows' is sorted, then we can write logic to move values from one reader
+     * at a time, identifying the rows in sequence that belong to the same block. This logic is more complex, but may
+     * perform better in the sorted case.
+     *
+     * An alternative is to not have multiple buffers, but just copy the values from one buffer to next as we
+     * increase the number of rows.
+     */
+    if (readers.size() == 1) {
+      readers.get(0).readIntValues(rows, 0, rowStartPos, rowSize, values, valuesStartPos);
+    }
+    int numRows = 0;
+    for (int rowIter = rowStartPos, valueIter = valuesStartPos; numRows < rowSize; numRows++, rowIter++, valueIter++) {
+      int row = rows[rowIter];
```

Comment:
This avoids some div and modulo costs but not sure if it's more optimal (so optional suggestion)
prevBufferIndex = getBufferId(rows[0]);
startRowIndex = (prevBufferIndex) * numRowsInChunk
endIndex = startRowIndex + numRowsInChunk
for() {
   if rows[rowIter] >= startRowIndex && rows[rowIter] < endRowIndex {
       values[valueIter] = readers.get(prevBufferIndex).getInt(endRowIndex - rows[rowsIter]);
    } else {
      prevBufferIndex = getBufferId(rows[0]);
      startRowIndex = (prevBufferIndex) * numRowsInChunk
      endIndex = startRowIndex + numRowsInChunk
      values[valueIter] =  readers.get(prevBufferIndex).getInt(endRowIndex - rows[rowsIter]);
    }
}

}